Date: Mon, 30 Oct 2023 19:03:38 +0000
From: Deekshit Devadas <deekshit.devadas@sky.uk>
Subject: [PATCH] RDK-42648: Idle CPU Time metrics V2
Idle Time metrics V2 to include - sum idel time, maxt idle time,
all cpu running time & it's sum , max runtime etc.
Source: d82de9e6772f4d7c92a0164f3accff8cd5fa388d Mon Sep 17 00:00:00 2001
Signed-off-by: deekshit.devadas <deekshit.devadas@sky.uk>
---

diff -Naur a/linux-4.9/include/uapi/linux/prctl.h b/linux-4.9/include/uapi/linux/prctl.h
--- a/include/uapi/linux/prctl.h	2023-10-30 19:03:38.551453448 +0000
+++ b/include/uapi/linux/prctl.h	2023-10-30 19:03:50.587452782 +0000
@@ -330,6 +330,58 @@
 	unsigned long long idle_time[T962X3_NUM_CPUS];
 } IDLE_METRICS;
 
+typedef struct {
+	unsigned long long sum_idle_time;
+	unsigned long max_idle_time;
+	unsigned long idle_count;
+} PER_CPU_IDLE_METRIC;
+
+/* define structure(s) used for logging idle metrics */
+typedef struct {
+	unsigned long long metricStartTime;
+	unsigned long long metricEndTime;
+	PER_CPU_IDLE_METRIC idle[T962X3_NUM_CPUS];
+	unsigned long long sum_all_cpus_running_time;
+	unsigned long long sum_run_time_lt_1ms;
+	unsigned long long sum_run_time_gt_1ms;
+	unsigned long long sum_run_time_gt_5ms;
+	unsigned long long sum_run_time_gt_10ms;
+	unsigned long long sum_run_time_gt_20ms;
+	unsigned long long sum_run_time_gt_30ms;
+	unsigned long long sum_run_time_gt_40ms;
+	unsigned long long sum_run_time_gt_50ms;
+	unsigned long long sum_run_time_gt_75ms;
+	unsigned long long sum_run_time_gt_100ms;
+	unsigned long max_run_time_lt_1ms;
+	unsigned long max_run_time_gt_1ms;
+	unsigned long max_run_time_gt_5ms;
+	unsigned long max_run_time_gt_10ms;
+	unsigned long max_run_time_gt_20ms;
+	unsigned long max_run_time_gt_30ms;
+	unsigned long max_run_time_gt_40ms;
+	unsigned long max_run_time_gt_50ms;
+	unsigned long max_run_time_gt_75ms;
+	unsigned long max_run_time_gt_100ms;
+	unsigned long max_all_cpus_running_time;
+	unsigned long count;
+	unsigned long run_time_gt_1ms;
+	unsigned long run_time_gt_5ms;
+	unsigned long run_time_gt_10ms;
+	unsigned long run_time_gt_20ms;
+	unsigned long run_time_gt_30ms;
+	unsigned long run_time_gt_40ms;
+	unsigned long run_time_gt_50ms;
+	unsigned long run_time_gt_75ms;
+	unsigned long run_time_gt_100ms;
+} IDLE_METRICS_V2;
+
 #define PR_GET_IDLE_METRICS		104
+#define PR_GET_IDLE_METRICS_VERSION	105
+
+typedef enum {
+	IDLE_METRICS_VERSION_V1 = 0,
+	IDLE_METRICS_VERSION_V2,
+	IDLE_METRICS_VERSION_NUM_VERSIONS
+} IDLE_METRICS_VERSION;
 
 #endif /* _LINUX_PRCTL_H */
